sqlglot/optimizer/canonicalize.py

```diff
@@ -22,7 +22,7 @@ def canonicalize(expression: exp.Expression, dialect: DialectType = None) -> exp
     dialect = Dialect.get_or_raise(dialect)
 
     def _canonicalize(expression: exp.Expression) -> exp.Expression:
-        expression = add_text_to_concat(expression)
+        expression = add_text_to_concat(expression, dialect.CONCAT_COALESCE)
```

Comment:
Worth testing this, e.g. canonicalize `select a::text + b::text` with `read="tsql"` and then transpile it to postgres. The result should use `||`:

```
1> select 'a' + null -- in tsql
2> go

--
NULL
```

```
(sqlglot) ➜  sqlglot git:(main) psql -c "select concat('a', null)"
 concat
--------
 a
(1 row)
```

Note: you need to annotate types before canonicalizing to hit this code path.